File: tradeit-android-sdk/src/main/java/it/trade/android/sdk/model/TradeItLinkedBrokerAccountParcelable.java
package it.trade.android.sdk.model;


import android.os.Parcel;
import android.os.Parcelable;

import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import it.trade.android.sdk.model.orderstatus.TradeItOrderStatusParcelable;
import it.trade.api.TradeItApiClient;
import it.trade.model.TradeItErrorResult;
import it.trade.model.callback.TradeItCallback;
import it.trade.model.reponse.Instrument;
import it.trade.model.reponse.OrderStatusDetails;
import it.trade.model.reponse.TradeItAccountOverviewResponse;
import it.trade.model.reponse.TradeItBrokerAccount;
import it.trade.model.reponse.TradeItPosition;

import static it.trade.model.reponse.TradeItErrorCode.BROKER_EXECUTION_ERROR;
import static it.trade.model.reponse.TradeItErrorCode.PARAMETER_ERROR;

public class TradeItLinkedBrokerAccountParcelable implements Parcelable {
    private static Map<String, TradeItLinkedBrokerParcelable> linkedBrokersMap = new HashMap<>(); //used for parcelable

    @SerializedName("accountName")
    protected String accountName;

    @SerializedName("accountNumber")
    protected String accountNumber;

    @SerializedName("accountBaseCurrency")
    protected String accountBaseCurrency;

    @SerializedName("userCanDisableMargin")
    protected boolean userCanDisableMargin;

    @SerializedName("orderCapabilities")
    protected List<TradeItOrderCapabilityParcelable> orderCapabilities;

    public transient TradeItLinkedBrokerParcelable linkedBroker;

    @SerializedName("balance")
    private TradeItBalanceParcelable balance;

    @SerializedName("fxBalance")
    private TradeItFxBalanceParcelable fxBalance;

    @SerializedName("balanceLastUpdated")
    private Date balanceLastUpdated;

    @SerializedName("positions")
    private List<TradeItPositionParcelable> positions;

    @SerializedName("ordersStatus")
    private List<TradeItOrderStatusParcelable> ordersStatus;

    @SerializedName("userId")
    private String userId;

    public TradeItLinkedBrokerAccountParcelable(TradeItLinkedBrokerParcelable linkedBroker, TradeItBrokerAccount account) {
        this.linkedBroker =  linkedBroker;
        this.accountName = account.name;
        this.accountNumber = account.accountNumber;
        this.accountBaseCurrency = account.accountBaseCurrency;
        this.userCanDisableMargin = account.userCanDisableMargin;
        this.orderCapabilities = TradeItOrderCapabilityParcelable.mapOrderCapabilitiesToTradeItOrderCapabilityParcelables(account.orderCapabilities);
        this.userId = linkedBroker.getLinkedLogin().userId;
    }

    protected TradeItApiClient getTradeItApiClient() {
        return this.linkedBroker.getApiClient();
    }

    protected void setErrorOnLinkedBroker(TradeItErrorResultParcelable errorResult) {
        if (errorResult.getErrorCode() != BROKER_EXECUTION_ERROR && errorResult.getErrorCode() != PARAMETER_ERROR) {
            this.linkedBroker.setError(errorResult);
        }
    }

    public String getBrokerName() {
        return this.linkedBroker.getBrokerName();
    }

    public String getAccountName() {
        return accountName;
    }

    public String getAccountNumber() {
        return accountNumber;
    }

    public String getAccountBaseCurrency() {
        return accountBaseCurrency;
    }

    public boolean getUserCanDisableMargin() {
        return userCanDisableMargin;
    }

    public void setUserCanDisableMargin(boolean userCanDisableMargin) {
        this.userCanDisableMargin = userCanDisableMargin;
    }

    public TradeItBalanceParcelable getBalance() {
        return balance;
    }

    public void setBalance(TradeItBalanceParcelable balance) {
        this.balance = balance;
    }

    public TradeItFxBalanceParcelable getFxBalance() {
        return fxBalance;
    }

    public void setFxBalance(TradeItFxBalanceParcelable fxBalance) {
        this.fxBalance = fxBalance;
    }

    public Date getBalanceLastUpdated() { return this.balanceLastUpdated; }

    public void setBalanceLastUpdated(Date balanceLastUpdated) {
        this.balanceLastUpdated = balanceLastUpdated;
    }

    public List<TradeItPositionParcelable> getPositions() {
        return positions;
    }

    public List<TradeItOrderStatusParcelable> getOrdersStatus() {
        return ordersStatus;
    }

    void setLinkedBroker(TradeItLinkedBrokerParcelable linkedBroker) {
        this.linkedBroker = linkedBroker;
    }

    void setPositions(List<TradeItPositionParcelable> positions) {
        this.positions = positions;
    }

    public TradeItOrderCapabilityParcelable getOrderCapabilityForInstrument(Instrument instrument) {
        for (TradeItOrderCapabilityParcelable orderCapability: orderCapabilities) {
            if (orderCapability.getInstrument() == instrument) {
                return orderCapability;
            }
        }
        return null;
    }

    public void refreshBalance(final TradeItCallback<TradeItLinkedBrokerAccountParcelable> callback) {
        final TradeItLinkedBrokerAccountParcelable linkedBrokerAccount = this;
        this.getTradeItApiClient().getAccountOverview(accountNumber, new TradeItCallback<TradeItAccountOverviewResponse>() {
            @Override
            public void onSuccess(TradeItAccountOverviewResponse response) {
                if (response.accountOverview != null) {
                    TradeItBalanceParcelable balance = new TradeItBalanceParcelable(response.accountOverview);
                    linkedBrokerAccount.balance = balance;
                    linkedBrokerAccount.fxBalance = null;
                } else if (response.fxAccountOverview != null) {
                    TradeItFxBalanceParcelable fxBalance = new TradeItFxBalanceParcelable(response.fxAccountOverview);
                    linkedBrokerAccount.balance = null;
                    linkedBrokerAccount.fxBalance = fxBalance;
                }
                linkedBrokerAccount.balanceLastUpdated = new Date();
                linkedBroker.cache();
                callback.onSuccess(linkedBrokerAccount);
            }

            @Override
            public void onError(TradeItErrorResult errorResult) {
                TradeItErrorResultParcelable errorResultParcelable = new TradeItErrorResultParcelable(errorResult);
                linkedBrokerAccount.setErrorOnLinkedBroker(errorResultParcelable);
                callback.onError(errorResultParcelable);
            }
        });
    }

    public void refreshPositions(final TradeItCallback<List<TradeItPositionParcelable>> callback) {
        final TradeItLinkedBrokerAccountParcelable linkedBrokerAccount = this;
        this.getTradeItApiClient().getPositions(accountNumber, new TradeItCallback<List<TradeItPosition>>() {
            @Override
            public void onSuccess(List<TradeItPosition> positions) {
                List<TradeItPositionParcelable> positionsParcelable = mapPositionsToPositionsParcelable(positions);
                linkedBrokerAccount.positions = positionsParcelable;
                callback.onSuccess(positionsParcelable);
            }

            @Override
            public void onError(TradeItErrorResult errorResult) {
                TradeItErrorResultParcelable errorResultParcelable = new TradeItErrorResultParcelable(errorResult);
                linkedBrokerAccount.setErrorOnLinkedBroker(errorResultParcelable);
                callback.onError(errorResultParcelable);
            }
        });
    }

    public void refreshOrdersStatus(final TradeItCallback<List<TradeItOrderStatusParcelable>> callback) {
        final TradeItLinkedBrokerAccountParcelable linkedBrokerAccount = this;
        this.getTradeItApiClient().getAllOrderStatus(accountNumber, new TradeItCallback<List<OrderStatusDetails>>() {
            @Override
            public void onSuccess(List<OrderStatusDetails> orderStatusDetailsList) {
                List<TradeItOrderStatusParcelable> orderStatusParcelables = TradeItOrderStatusParcelable.mapOrderStatusDetailsToTradeItOrderStatusParcelable(orderStatusDetailsList);
                linkedBrokerAccount.ordersStatus = orderStatusParcelables;
                callback.onSuccess(orderStatusParcelables);
            }

            @Override
            public void onError(TradeItErrorResult error) {
                TradeItErrorResultParcelable errorResultParcelable = new TradeItErrorResultParcelable(error);
                linkedBrokerAccount.setErrorOnLinkedBroker(errorResultParcelable);
                callback.onError(errorResultParcelable);
            }
        });
    }

    public void cancelOrder(String orderNumber, final TradeItCallback<TradeItOrderStatusParcelable> callback) {
        final TradeItLinkedBrokerAccountParcelable linkedBrokerAccount = this;
        this.getTradeItApiClient().cancelOrder(accountNumber, orderNumber, new TradeItCallback<OrderStatusDetails>() {
            @Override
            public void onSuccess(OrderStatusDetails orderStatusDetails) {
                callback.onSuccess(new TradeItOrderStatusParcelable(orderStatusDetails));
            }

            @Override
            public void onError(TradeItErrorResult error) {
                TradeItErrorResultParcelable errorResultParcelable = new TradeItErrorResultParcelable(error);
                linkedBrokerAccount.setErrorOnLinkedBroker(errorResultParcelable);
                callback.onError(errorResultParcelable);
            }
        });
    }

    private List<TradeItPositionParcelable> mapPositionsToPositionsParcelable(List<TradeItPosition> positions) {
        List<TradeItPositionParcelable> positionsParcelable = new ArrayList<>();
        for (TradeItPosition position : positions) {
            positionsParcelable.add(new TradeItPositionParcelable(position));
        }
        return positionsParcelable;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        TradeItLinkedBrokerAccountParcelable that = (TradeItLinkedBrokerAccountParcelable) o;

        if (accountName != null ? !accountName.equals(that.accountName) : that.accountName != null)
            return false;
        if (accountNumber != null ? !accountNumber.equals(that.accountNumber) : that.accountNumber != null)
            return false;
        return accountBaseCurrency != null ? accountBaseCurrency.equals(that.accountBaseCurrency) : that.accountBaseCurrency == null;

    }

    @Override
    public int hashCode() {
        int result = accountName != null ? accountName.hashCode() : 0;
        result = 31 * result + (accountNumber != null ? accountNumber.hashCode() : 0);
        result = 31 * result + (accountBaseCurrency != null ? accountBaseCurrency.hashCode() : 0);
        return result;
    }

    @Override
    public String toString() {
        return "TradeItLinkedBrokerAccountParcelable{" +
                "accountBaseCurrency='" + accountBaseCurrency + '\'' +
                ", accountName='" + accountName + '\'' +
                ", accountNumber='" + accountNumber + '\'' +
                ", userCanDisableMargin='" + userCanDisableMargin + '\'' +
                ", balance='" + balance + '\'' +
                ", orderCapabilities='" + orderCapabilities + '\'' +
                '}';
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.accountName);
        dest.writeString(this.accountNumber);
        dest.writeString(this.accountBaseCurrency);
        dest.writeByte((byte) (userCanDisableMargin ? 1 : 0));
        dest.writeList(this.orderCapabilities);
        dest.writeParcelable(this.balance, flags);
        dest.writeLong(this.balanceLastUpdated != null ? this.balanceLastUpdated.getTime() : -1);
        dest.writeList(this.positions);
        dest.writeString(this.userId);
        linkedBrokersMap.put(this.userId, this.linkedBroker);
    }

    protected TradeItLinkedBrokerAccountParcelable(Parcel in) {
        this.accountName = in.readString();
        this.accountNumber = in.readString();
        this.accountBaseCurrency = in.readString();
        this.userCanDisableMargin = in.readByte() != 0;
        this.orderCapabilities = new ArrayList<>();
        in.readList(this.orderCapabilities, TradeItOrderCapabilityParcelable.class.getClassLoader());
        this.balance = in.readParcelable(TradeItBalanceParcelable.class.getClassLoader());
        long tmpBalanceLastUpdated = in.readLong();
        this.balanceLastUpdated = tmpBalanceLastUpdated == -1 ? null : new Date(tmpBalanceLastUpdated);
        this.positions = new ArrayList<TradeItPositionParcelable>();
        in.readList(this.positions, TradeItPositionParcelable.class.getClassLoader());
        this.userId = in.readString();
        this.linkedBroker = linkedBrokersMap.get(this.userId);
        int indexAccount = this.linkedBroker.getAccounts().indexOf(this);
        if (indexAccount != -1) { // updating account reference on the linkedBroker as we created a new object
            this.linkedBroker.getAccounts().remove(indexAccount);
            this.linkedBroker.getAccounts().add(indexAccount, this);
        }
    }

    public static final Parcelable.Creator<TradeItLinkedBrokerAccountParcelable> CREATOR = new Parcelable.Creator<TradeItLinkedBrokerAccountParcelable>() {
        @Override
        public TradeItLinkedBrokerAccountParcelable createFromParcel(Parcel source) {
            return new TradeItLinkedBrokerAccountParcelable(source);
        }

        @Override
        public TradeItLinkedBrokerAccountParcelable[] newArray(int size) {
            return new TradeItLinkedBrokerAccountParcelable[size];
        }
    };
}

File: tradeit-android-sdk/src/main/kotlin/it/trade/android/sdk/model/TradeItLinkedBrokerAccountParcelable.kt
package it.trade.android.sdk.model


import android.os.Parcel
import android.os.Parcelable
import com.google.gson.annotations.SerializedName
import it.trade.android.sdk.model.orderstatus.TradeItOrderStatusParcelable
import it.trade.api.TradeItApiClient
import it.trade.model.TradeItErrorResult
import it.trade.model.callback.TradeItCallback
import it.trade.model.reponse.*
import it.trade.model.reponse.TradeItErrorCode.BROKER_EXECUTION_ERROR
import it.trade.model.reponse.TradeItErrorCode.PARAMETER_ERROR
import java.util.ArrayList
import java.util.Date
import kotlin.collections.HashMap

class TradeItLinkedBrokerAccountParcelable : Parcelable {

    @SerializedName("accountName")
    var accountName: String

    @SerializedName("accountNumber")
    var accountNumber: String

    @SerializedName("accountBaseCurrency")
    var accountBaseCurrency: String

    @SerializedName("accountIndex")
    var accountIndex: String? = ""

    @SerializedName("tradable")
    var tradable: Boolean = true

    @SerializedName("userCanDisableMargin")
    var userCanDisableMargin: Boolean = false

    @SerializedName("orderCapabilities")
    var orderCapabilities: List<TradeItOrderCapabilityParcelable>

    @Transient
    var linkedBroker: TradeItLinkedBrokerParcelable?
        internal set

    @SerializedName("balance")
    var balance: TradeItBalanceParcelable? = null

    @SerializedName("fxBalance")
    var fxBalance: TradeItFxBalanceParcelable? = null

    @SerializedName("balanceLastUpdated")
    var balanceLastUpdated: Date? = null

    @SerializedName("positions")
    var positions: List<TradeItPositionParcelable>? = null
        internal set

    @SerializedName("ordersStatus")
    var ordersStatus: List<TradeItOrderStatusParcelable>? = null
        private set

    @SerializedName("userId")
    private var userId: String? = null

    val tradeItApiClient: TradeItApiClient?
        get() = this.linkedBroker?.getApiClient()


    val brokerName: String
        get() = this.linkedBroker?.brokerName ?: ""

    constructor(linkedBroker: TradeItLinkedBrokerParcelable, account: TradeItBrokerAccount) {
        this.linkedBroker = linkedBroker
        this.accountName = account.name
        this.accountNumber = account.accountNumber
        this.accountIndex = account.accountIndex
        this.tradable = account.tradable
        this.accountBaseCurrency = account.accountBaseCurrency
        this.userCanDisableMargin = account.userCanDisableMargin
        this.orderCapabilities = TradeItOrderCapabilityParcelable.mapOrderCapabilitiesToTradeItOrderCapabilityParcelables(account.orderCapabilities)
        this.userId = linkedBroker.linkedLogin!!.userId
    }

    fun setErrorOnLinkedBroker(errorResult: TradeItErrorResultParcelable) {
        if (errorResult.errorCode != BROKER_EXECUTION_ERROR && errorResult.errorCode != PARAMETER_ERROR) {
            this.linkedBroker?.error = errorResult
        }
    }

    fun getOrderCapabilityForInstrument(instrument: Instrument): TradeItOrderCapabilityParcelable? {
        for (orderCapability in orderCapabilities) {
            if (orderCapability.instrument == instrument) {
                return orderCapability
            }
        }
        return null
    }

    fun refreshBalance(callback: TradeItCallback<TradeItLinkedBrokerAccountParcelable>) {
        val linkedBrokerAccount = this
        this.tradeItApiClient!!.getAccountOverview(accountNumber, object : TradeItCallback<TradeItAccountOverviewResponse> {
            override fun onSuccess(response: TradeItAccountOverviewResponse) {
                if (response.accountOverview != null) {
                    val balance = TradeItBalanceParcelable(response.accountOverview)
                    linkedBrokerAccount.balance = balance
                    linkedBrokerAccount.fxBalance = null
                } else if (response.fxAccountOverview != null) {
                    val fxBalance = TradeItFxBalanceParcelable(response.fxAccountOverview)
                    linkedBrokerAccount.balance = null
                    linkedBrokerAccount.fxBalance = fxBalance
                }
                linkedBrokerAccount.balanceLastUpdated = Date()
                linkedBroker?.cache()
                callback.onSuccess(linkedBrokerAccount)
            }

            override fun onError(errorResult: TradeItErrorResult) {
                val errorResultParcelable = TradeItErrorResultParcelable(errorResult)
                linkedBrokerAccount.setErrorOnLinkedBroker(errorResultParcelable)
                callback.onError(errorResultParcelable)
            }
        })
    }

    fun refreshPositions(callback: TradeItCallback<List<TradeItPositionParcelable>>) {
        val linkedBrokerAccount = this
        this.tradeItApiClient!!.getPositions(accountNumber, object : TradeItCallback<List<TradeItPosition>> {
            override fun onSuccess(positions: List<TradeItPosition>) {
                val positionsParcelable = mapPositionsToPositionsParcelable(positions)
                linkedBrokerAccount.positions = positionsParcelable
                callback.onSuccess(positionsParcelable)
            }

            override fun onError(errorResult: TradeItErrorResult) {
                val errorResultParcelable = TradeItErrorResultParcelable(errorResult)
                linkedBrokerAccount.setErrorOnLinkedBroker(errorResultParcelable)
                callback.onError(errorResultParcelable)
            }
        })
    }

    fun refreshOrdersStatus(callback: TradeItCallback<List<TradeItOrderStatusParcelable>>) {
        val linkedBrokerAccount = this
        this.tradeItApiClient!!.getAllOrderStatus(accountNumber, object : TradeItCallback<List<OrderStatusDetails>> {
            override fun onSuccess(orderStatusDetailsList: List<OrderStatusDetails>) {
                val orderStatusParcelables = TradeItOrderStatusParcelable.mapOrderStatusDetailsToTradeItOrderStatusParcelable(orderStatusDetailsList)
                linkedBrokerAccount.ordersStatus = orderStatusParcelables
                callback.onSuccess(orderStatusParcelables)
            }

            override fun onError(error: TradeItErrorResult) {
                val errorResultParcelable = TradeItErrorResultParcelable(error)
                linkedBrokerAccount.setErrorOnLinkedBroker(errorResultParcelable)
                callback.onError(errorResultParcelable)
            }
        })
    }

    fun cancelOrder(orderNumber: String, callback: TradeItCallback<TradeItOrderStatusParcelable>) {
        val linkedBrokerAccount = this
        this.tradeItApiClient!!.cancelOrder(accountNumber, orderNumber, object : TradeItCallback<OrderStatusDetails> {
            override fun onSuccess(orderStatusDetails: OrderStatusDetails) {
                callback.onSuccess(TradeItOrderStatusParcelable(orderStatusDetails))
            }

            override fun onError(error: TradeItErrorResult) {
                val errorResultParcelable = TradeItErrorResultParcelable(error)
                linkedBrokerAccount.setErrorOnLinkedBroker(errorResultParcelable)
                callback.onError(errorResultParcelable)
            }
        })
    }

    fun getCryptoQuote(pair: String, callback: TradeItCallback<TradeItCryptoQuoteResponseParcelable>) {
        val linkedBrokerAccount = this
        this.tradeItApiClient?.getCryptoQuote(
            this.accountNumber,
            pair,
            object: TradeItCallback<TradeItCryptoQuoteResponse> {
                override fun onSuccess(response: TradeItCryptoQuoteResponse) {
                    callback.onSuccess(TradeItCryptoQuoteResponseParcelable(response))
                }

                override fun onError(error: TradeItErrorResult) {
                    val errorResultParcelable = TradeItErrorResultParcelable(error)
                    linkedBrokerAccount.setErrorOnLinkedBroker(errorResultParcelable)
                    callback.onError(errorResultParcelable)
                }

            }
        )
    }

    fun getProxyVoteUrl(symbol: String, callback: TradeItCallback<String>) {
        this.tradeItApiClient?.getProxyVoteUrl(
            accountNumber,
            symbol,
            object : TradeItCallback<TradeItProxyVoteUrlResponse> {
                override fun onSuccess(response: TradeItProxyVoteUrlResponse) {
                    response.proxyVoteUrl?.let {  callback.onSuccess(it)}
                        ?: callback.onError(TradeItErrorResultParcelable(
                            TradeItErrorCode.SYSTEM_ERROR,
                            "Position not eligible",
                            arrayListOf("This position is not eligible to proxy voting")
                        ))
                }

                override fun onError(error: TradeItErrorResult) {
                    val errorResultParcelable = TradeItErrorResultParcelable(error)
                    callback.onError(errorResultParcelable)
                }
            }
        )
    }

    private fun mapPositionsToPositionsParcelable(positions: List<TradeItPosition>): List<TradeItPositionParcelable> {
        val positionsParcelable = ArrayList<TradeItPositionParcelable>()
        for (position in positions) {
            positionsParcelable.add(TradeItPositionParcelable(position))
        }
        return positionsParcelable
    }


    override fun toString(): String {
        return "TradeItLinkedBrokerAccountParcelable{" +
                "accountBaseCurrency='" + accountBaseCurrency + '\''.toString() +
                ", accountName='" + accountName + '\''.toString() +
                ", accountNumber='" + accountNumber + '\''.toString() +
                ", accountIndex='" + accountIndex + '\''.toString() +
                ", tradable='" + tradable + '\''.toString() +
                ", userCanDisableMargin='" + userCanDisableMargin + '\''.toString() +
                ", balance='" + balance + '\''.toString() +
                ", orderCapabilities='" + orderCapabilities + '\''.toString() +
                '}'.toString()
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(this.accountName)
        dest.writeString(this.accountNumber)
        dest.writeString(this.accountBaseCurrency)
        dest.writeByte((if (userCanDisableMargin) 1 else 0).toByte())
        dest.writeList(this.orderCapabilities)
        dest.writeParcelable(this.balance, flags)
        dest.writeLong(if (this.balanceLastUpdated != null) this.balanceLastUpdated!!.time else -1)
        dest.writeList(this.positions)
        dest.writeString(this.userId)
        userId?.let {userId ->  linkedBrokersMap[userId] = this.linkedBroker }
        dest.writeString(this.accountIndex)
        dest.writeByte((if (tradable) 1 else 0).toByte())
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as TradeItLinkedBrokerAccountParcelable

        if (accountName != other.accountName) return false
        if (accountNumber != other.accountNumber) return false
        if (accountBaseCurrency != other.accountBaseCurrency) return false

        return true
    }

    override fun hashCode(): Int {
        var result = accountName.hashCode()
        result = 31 * result + accountNumber.hashCode()
        result = 31 * result + accountBaseCurrency.hashCode()
        result = 31 * result + (accountIndex?.hashCode() ?: 0)
        result = 31 * result + tradable.hashCode()
        result = 31 * result + userCanDisableMargin.hashCode()
        result = 31 * result + orderCapabilities.hashCode()
        result = 31 * result + (linkedBroker?.hashCode() ?: 0)
        result = 31 * result + (balance?.hashCode() ?: 0)
        result = 31 * result + (fxBalance?.hashCode() ?: 0)
        result = 31 * result + (balanceLastUpdated?.hashCode() ?: 0)
        result = 31 * result + (positions?.hashCode() ?: 0)
        result = 31 * result + (ordersStatus?.hashCode() ?: 0)
        result = 31 * result + (userId?.hashCode() ?: 0)
        return result
    }

    protected constructor(`in`: Parcel) {
        this.accountName = `in`.readString()
        this.accountNumber = `in`.readString()
        this.accountBaseCurrency = `in`.readString()
        this.userCanDisableMargin = `in`.readByte().toInt() != 0
        this.orderCapabilities = ArrayList()
        `in`.readList(this.orderCapabilities, TradeItOrderCapabilityParcelable::class.java.getClassLoader())
        this.balance = `in`.readParcelable(TradeItBalanceParcelable::class.java.getClassLoader())
        val tmpBalanceLastUpdated = `in`.readLong()
        this.balanceLastUpdated = if (tmpBalanceLastUpdated.equals(-1)) null else Date(tmpBalanceLastUpdated)
        this.positions = ArrayList()
        `in`.readList(this.positions, TradeItPositionParcelable::class.java.getClassLoader())
        this.userId = `in`.readString()
        this.linkedBroker = this.userId?.let { userId -> linkedBrokersMap[userId] }
        this.linkedBroker?.accounts?.indexOf(this)?.let { indexAccount ->
            if (indexAccount != -1) { // updating account reference on the linkedBroker as we created a new object
                this.linkedBroker?.accounts?.removeAt(indexAccount)
                this.linkedBroker?.accounts?.add(indexAccount, this)
            }
        }
        this.accountIndex = `in`.readString()
        this.tradable = `in`.readByte().toInt() != 0
    }

    companion object {
        private val linkedBrokersMap = HashMap<String, TradeItLinkedBrokerParcelable?>() //used for parcelable
        @JvmField
        val CREATOR: Parcelable.Creator<TradeItLinkedBrokerAccountParcelable> = object : Parcelable.Creator<TradeItLinkedBrokerAccountParcelable> {
            override fun createFromParcel(source: Parcel): TradeItLinkedBrokerAccountParcelable {
                return TradeItLinkedBrokerAccountParcelable(source)
            }

            override fun newArray(size: Int): Array<TradeItLinkedBrokerAccountParcelable?> {
                return arrayOfNulls(size)
            }
        }
    }
}
